### PR TITLE
Kanu UI 246 adding addpushsource event from UI

### PR DESCRIPTION
**Pull request by dmitriy-borzenko**

### CODE REVIEW

File: src/app/api/gql/fragments/fragment-metadata-block.graphql [hunk 1]
```diff
@@ -18,5 +18,8 @@ fragment MetadataBlock on MetadataBlockExtended {
     ...DatasetCurrentInfo
     ...SetLicenseEvent
     ...SetPollingSourceEvent
+    ...AddPushSourceEvent
+    ...SetDataSchemaEvent
+    ...DisablePollingSourceEvent
```
Comment (zaychenko-sergei) [inline src/app/api/gql/fragments/fragment-metadata-block.graphql:23]: Disable push source too?

File: src/app/api/gql/fragments/events/fragment-add-push-source.graphql [hunk 2]
```diff
@@ -0,0 +1,93 @@
+fragment AddPushSourceEvent on AddPushSource {
+  sourceName
+  read {
```
Comment (zaychenko-sergei) [inline src/app/api/gql/fragments/events/fragment-add-push-source.graphql:3]: There is large duplication with fragment for `SetPollingSource`. Extract common definitions for steps and reuse in both events.

File: src/app/dataset-block/metadata-block/components/event-details/components/add-data-event/add-data-event.source.ts [hunk 3]
```diff
@@ -4,11 +4,12 @@ import { HashPropertyComponent } from "../common/hash-property/hash-property.com
 import { OffsetIntervalPropertyComponent } from "../common/offset-interval-property/offset-interval-property.component";
 import { SizePropertyComponent } from "../common/size-property/size-property.component";
 import { AddDataToolipsTexts } from "src/app/common/tooltips/add-data.text";
+import { AddPushSourceToolipsTexts } from "src/app/common/tooltips/add-push-source.text";
 
 export const ADD_DATA_SOURCE_DESCRIPTORS: EventRowDescriptorsByField = {
     "AddData.string.addDataWatermark": {
         label: "Watermark:",
-        tooltip: AddDataToolipsTexts.WATERMARK,
+        tooltip: AddPushSourceToolipsTexts.SOURCE_NAME,
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/add-data-event/add-data-event.source.ts:12]: Probably a mistake

File: src/app/dataset-block/metadata-block/components/event-details/components/add-push-source-event/add-push-source-event.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import { ComponentFixture, TestBed } from "@angular/core/testing";
+import { AddPushSourceEventComponent } from "./add-push-source-event.component";
+import { mockAddPushSource } from "../../mock.events";
+import { SharedTestModule } from "src/app/common/shared-test.module";
+import { BlockRowDataComponent } from "../../../block-row-data/block-row-data.component";
+import { TooltipIconComponent } from "../../../tooltip-icon/tooltip-icon.component";
+import { MatIconModule } from "@angular/material/icon";
+import { NgbTooltipModule } from "@ng-bootstrap/ng-bootstrap";
+
+describe("AddPushSourceEventComponent", () => {
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/add-push-source-event/add-push-source-event.component.spec.ts:10]: Add some tests

File: src/app/dataset-block/metadata-block/components/event-details/components/add-push-source-event/add-push-source-event.source.ts [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+        label: "Source name:",
+        tooltip: AddPushSourceToolipsTexts.SOURCE_NAME,
+        presentationComponent: SimplePropertyComponent,
+        separateRowForValue: false,
+        dataTestId: "add-push-source-name",
+    },
+
+    "AddPushSource.ReadStepCsv.__typename": {
+        label: "Type:",
+        tooltip: SetPollingSourceToolipsTexts.READ_CSV,
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/add-push-source-event/add-push-source-event.source.ts:24]: Rather move shared tooltips to a separate file and use in both events.

File: src/app/dataset-block/metadata-block/components/event-details/components/add-push-source-event/add-push-source-event.source.ts [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+        presentationComponent: StepTypePropertyComponent,
+        separateRowForValue: false,
+        dataTestId: "add-push-source-ReadStepCsv-__typename",
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/add-push-source-event/add-push-source-event.source.ts:27]: For identical properties, if the only difference is prefix in `dataTestId`, we should have a mechanism to have a single definition with prefix given as a parameter.

File: src/app/dataset-block/metadata-block/components/event-details/event-details.component.html [hunk 7]
```diff
@@ -39,6 +39,14 @@
         <app-set-watermark-event [event]="setWatermarkEvent"></app-set-watermark-event>
     </ng-container>
 
+    <ng-container *ngIf="isAddPushSourceEvent">
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/event-details.component.html:42]: Can you remind why we are not using dynamic components to implement variation here? I remember some issues with clipboard copying.

File: src/app/dataset-view/additional-components/metadata-component/components/add-polling-source/add-polling-source.component.ts [hunk 8]
```diff
@@ -109,8 +109,8 @@ export class AddPollingSourceComponent extends BaseMainEventComponent implements
         this.subsribeErrorMessage();
     }
 
-    public changeStep(step: SetPollingSourceSection): void {
-        this.currentStep = step;
+    public changeStep(step: SetPollingSourceSection | AddPushSourceSection): void {
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/metadata-component/components/add-polling-source/add-polling-source.component.ts:112]: Make a type definition. This is used in many places

File: src/app/dataset-view/additional-components/metadata-component/components/add-polling-source/process-form.service.ts [hunk 9]
```diff
@@ -2,22 +2,25 @@ import { Injectable } from "@angular/core";
...
 })
 export class ProcessFormService {
     public transformForm(formGroup: FormGroup): void {
         this.transformSchema(formGroup);
-        this.processFetchOrderControl(formGroup);
+        if (has(formGroup.value, "fetch")) {
+            this.processFetchOrderControl(formGroup);
+            this.processEventTimeControl(formGroup);
+            this.processPipeCommandControl(formGroup);
+        }
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/metadata-component/components/add-polling-source/process-form.service.ts:21]: This correction is clearly to handle push sources, but the service itself is placed within `add-polling-source` folder.

File: src/app/dataset-view/additional-components/metadata-component/components/add-polling-source/process-form.service.ts [hunk 10]
```diff
@@ -65,30 +68,47 @@ export class ProcessFormService {
...
-            SetPollingSourceSection.MERGE,
-            SetPollingSourceSection.FETCH,
-        ];
-        formKeys.forEach((formKey: FormKeys) => {
-            Object.entries(form[formKey]).forEach(([key, value]) => {
-                if (!value || (Array.isArray(value) && !value.length)) {
-                    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-dynamic-delete
-                    delete formGroup.value[formKey][key];
-                }
+        if (has(formGroup.value, "fetch")) {
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/metadata-component/components/add-polling-source/process-form.service.ts:71]: Maybe early return with opposite condition?

File: src/app/dataset-view/additional-components/metadata-component/components/add-push-source/steps/source-name-step/source-name-step.component.html [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+        <app-input-field
+            [form]="form"
+            [controlName]="'sourceName'"
+            [label]="'Source name'"
+            [value]="''"
+            [placeholder]="'Enter name'"
+            [tooltip]="'Source name'"
+        ></app-input-field>
+        <div class="mt-1" *ngIf="form?.get('sourceName')?.errors?.noneOf">
+            <small class="text-danger ps-2"> {{ "Name is already exist" }}</small>
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/metadata-component/components/add-push-source/steps/source-name-step/source-name-step.component.html:14]: Name already exists

File: src/app/dataset-view/additional-components/overview-component/overview.component.ts [hunk 12]
```diff
@@ -110,6 +110,18 @@ export class OverviewComponent extends BaseComponent implements OnInit {
         }
     }
 
+    public get canAddPushSource(): boolean {
+        if (this.currentState && this.datasetPermissions.permissions.canCommit) {
+            return (
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/overview-component/overview.component.ts:115]: Place a TODO, so you don't forget to replace it to faster API

**Review by zaychenko-sergei — CHANGES_REQUESTED** [review]
I wonder if there is any approach to make `StepperNavigationComponent` generic over the source event type. The union does not look good: `SetPollingSourceSection | AddPushSourceSection` - would be cool to make the component, or at least it's base class, generic.

File: src/app/api/gql/fragments/fragment-metadata-block.graphql (hunk 1 above)
Comment (dmitriy-borzenko) [reply src/app/api/gql/fragments/fragment-metadata-block.graphql:23]: I'll add it when I do the metadata tab with editing

File: src/app/dataset-block/metadata-block/components/event-details/components/add-data-event/add-data-event.source.ts (hunk 3 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-block/metadata-block/components/event-details/components/add-data-event/add-data-event.source.ts:12]: Yes, overdid it )

File: src/app/dataset-block/metadata-block/components/event-details/components/add-push-source-event/add-push-source-event.source.ts (hunk 5 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-block/metadata-block/components/event-details/components/add-push-source-event/add-push-source-event.source.ts:24]: Moved shared tooltips to a separate file

File: src/app/dataset-view/additional-components/overview-component/overview.component.ts (hunk 12 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/additional-components/overview-component/overview.component.ts:115]: Add TODO: comments

File: src/app/dataset-view/additional-components/metadata-component/components/add-push-source/steps/source-name-step/source-name-step.component.html (hunk 11 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/additional-components/metadata-component/components/add-push-source/steps/source-name-step/source-name-step.component.html:14]: Fixed

File: src/app/dataset-block/metadata-block/components/event-details/event-details.component.html (hunk 7 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-block/metadata-block/components/event-details/event-details.component.html:42]: We solved the problem with clipboard copying. We installed a third party library for this

File: src/app/dataset-view/additional-components/metadata-component/components/add-polling-source/add-polling-source.component.ts (hunk 8 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/additional-components/metadata-component/components/add-polling-source/add-polling-source.component.ts:112]: Created  `type SourcesSection=SetPollingSourceSection | AddPushSourceSection`

File: src/app/dataset-view/additional-components/metadata-component/components/add-polling-source/process-form.service.ts (hunk 10 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/additional-components/metadata-component/components/add-polling-source/process-form.service.ts:71]: Fixed

File: src/app/dataset-view/additional-components/metadata-component/components/add-polling-source/process-form.service.ts (hunk 9 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-view/additional-components/metadata-component/components/add-polling-source/process-form.service.ts:21]: Moved to a separate folder

File: src/app/api/gql/fragments/events/fragment-add-push-source.graphql (hunk 2 above)
Comment (dmitriy-borzenko) [reply src/app/api/gql/fragments/events/fragment-add-push-source.graphql:3]: Fixed

File: src/app/dataset-block/metadata-block/components/event-details/components/add-push-source-event/add-push-source-event.component.spec.ts (hunk 4 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-block/metadata-block/components/event-details/components/add-push-source-event/add-push-source-event.component.spec.ts:10]: Added unit tests

File: src/app/dataset-block/metadata-block/components/event-details/components/set-transform-event/set-transform-event.source.ts [hunk 13]
```diff
@@ -6,23 +6,12 @@ import { OwnerPropertyComponent } from "../common/owner-property/owner-property.
 import { EventRowDescriptorsByField } from "../../dynamic-events/dynamic-events.model";
 import { DatasetNameByIdPropertyComponent } from "../common/dataset-name-by-id-property/dataset-name-by-id-property.component";
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/set-transform-event/set-transform-event.source.ts:7]: ESLint warnings are not for fun

File: src/app/dataset-block/metadata-block/components/event-details/components/unsupported-event/unsupported-event.component.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
+import { ChangeDetectionStrategy, Component } from "@angular/core";
+import { BaseComponent } from "src/app/common/base.component";
+
+@Component({
+    selector: "app-unsupported-event",
+    templateUrl: "./unsupported-event.component.html",
+
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/components/unsupported-event/unsupported-event.component.ts:7]: Useless blank

File: src/app/dataset-block/metadata-block/components/event-details/event-details.component.ts [hunk 15]
```diff
@@ -1,118 +1,116 @@
...
+                this.componentEventTypeFactory(this.currentBlock.event.__typename as SupportedEvents),
+            );
+            componentRef.setInput("event", this.currentBlock.event);
+            componentRef.changeDetectorRef.detectChanges();
+        }
+    }
+
+    private componentEventTypeFactory(type: SupportedEvents): Type<BaseComponent> {
+        let component: Type<BaseComponent>;
+        switch (type) {
```
Comment (zaychenko-sergei) [inline src/app/dataset-block/metadata-block/components/event-details/event-details.component.ts:58]: A syntax based on dictionary might be more compact than switch statement.

**Review by zaychenko-sergei — CHANGES_REQUESTED** [review]

File: src/app/dataset-block/metadata-block/components/event-details/components/unsupported-event/unsupported-event.component.ts (hunk 14 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-block/metadata-block/components/event-details/components/unsupported-event/unsupported-event.component.ts:7]: Fixed

File: src/app/dataset-block/metadata-block/components/event-details/components/set-transform-event/set-transform-event.source.ts (hunk 13 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-block/metadata-block/components/event-details/components/set-transform-event/set-transform-event.source.ts:7]: Fixed

File: src/app/dataset-block/metadata-block/components/event-details/event-details.component.ts (hunk 15 above)
Comment (dmitriy-borzenko) [reply src/app/dataset-block/metadata-block/components/event-details/event-details.component.ts:58]: Changed to dictionary

File: src/app/api/gql/fragments/events/fragment-add-push-source.graphql [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+    ...ReadStepCsvData
+    ...ReadStepGeoJsonData
+    ...ReadStepParquetData
+    ...ReadStepJsonLinesData
+    ...ReadStepEsriShapefileData
+    ...ReadStepJsonData
+    ...ReadStepNdGeoJsonData
+    ...ReadStepNdJsonData
```
[the remark this answers is not in the document]
Comment (dmitriy-borzenko) [reply src/app/api/gql/fragments/events/fragment-add-push-source.graphql:11]: Discussed this comment

File: src/app/dataset-view/additional-components/metadata-component/components/add-push-source/steps/source-name-step/source-name-step.component.html [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+        <app-input-field
+            [form]="form"
+            [controlName]="'sourceName'"
+            [label]="'Source name'"
+            [value]="''"
+            [placeholder]="'Enter name'"
+            [tooltip]="'Source name'"
+        ></app-input-field>
+        <div class="mt-1" *ngIf="form?.get('sourceName')?.errors?.noneOf">
+            <small class="text-danger ps-2"> {{ "Name already exist" }}</small>
```
Comment (zaychenko-sergei) [inline src/app/dataset-view/additional-components/metadata-component/components/add-push-source/steps/source-name-step/source-name-step.component.html:14]: exist**s**

**Review by zaychenko-sergei — APPROVED** [review]

**Pull request merged**